File: sub.py
import random
import requests
import threading
import time
from bs4 import BeautifulSoup
import re
from Crypto.Cipher import AES

def get_cookie():
    html = requests.get("https://codeforces.com").text
    str_a = re.search(r'a=toNumbers\("([a-f0-9]{32})"\)', html).group(1)
    str_b = re.search(r'b=toNumbers\("([a-f0-9]{32})"\)', html).group(1)
    str_c = re.search(r'c=toNumbers\("([a-f0-9]{32})"\)', html).group(1)

    a = bytes.fromhex(str_a)
    b = bytes.fromhex(str_b)
    c = bytes.fromhex(str_c)

    aes = AES.new(a, AES.MODE_CBC, iv=b)
    return {"RCPC": aes.decrypt(c).hex()}

def time_cache(tl):
    def inner(f):
        lock = threading.Lock()
        t = 0
        res = None

        def foo():
            nonlocal t, res

            with lock:
                t2 = time.time()
                if t2 > t:
                    res = f()
                    t = t2 + tl
                return res
        return foo
    return inner

@time_cache(3600)
def get_list():
    r = requests.get(f'https://codeforces.com/api/user.status?handle=TheScrasse')
    return r.json()['result']

def get_sub():
    r = random.choice(get_list())
    id, contest_id = r['id'], r['contestId']

    html = requests.get(f'https://codeforces.com/contest/{contest_id}/submission/{id}', cookies=get_cookie()).text
    soup = BeautifulSoup(html, "html.parser")
    problem = r['problem']

    problem_info = f"{problem['contestId']} {problem['index']} - {problem['name']} - *{r['verdict']}*"
    code = soup.find(id='program-source-text').text
    return problem_info + '\n' + f'`{code}`'

if __name__ == '__main__':
    print(get_sub())

File: src/sub.rs
use rand::seq::SliceRandom;
use scraper::{Html, Selector};
use serde::Deserialize;
use std::time::{Duration, Instant};
use teloxide::utils::markdown;
use tokio::sync::Mutex;

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct Sub {
    id: u64,
    contest_id: u64,
    problem: Problem,
    verdict: String,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct Problem {
    index: String,
    name: String,
}

async fn get_list() -> Result<Vec<Sub>, Box<dyn std::error::Error>> {
    #[derive(Deserialize)]
    struct Response {
        result: Vec<Sub>,
    }

    lazy_static::lazy_static! {
        static ref CACHE: Mutex<Option<(Instant, Vec<Sub>)>> = Mutex::new(None);
    }

    let mut lock = CACHE.lock().await;
    let now = Instant::now();

    if let Some((time, subs)) = &*lock {
        if now.duration_since(*time) < Duration::from_secs(3600 * 24) {
            return Ok(subs.clone());
        }
    }

    let url = "https://codeforces.com/api/user.status?handle=TheScrasse";
    let resp = reqwest::get(url).await?.json::<Response>().await?;
    let subs = resp.result;

    *lock = Some((now, subs.clone()));
    Ok(subs)
}

pub async fn get_sub() -> Result<String, Box<dyn std::error::Error>> {
    let list = get_list().await?;
    let Sub {
        id,
        contest_id,
        problem,
        verdict,
    } = list
        .choose(&mut rand::thread_rng())
        .ok_or("No submissions")?;

    log::info!("Getting sub {} from contest {}", id, contest_id);

    // let cookies = reqwest::CookieJar::new();

    let url = format!(
        "https://codeforces.com/contest/{}/submission/{}",
        contest_id, id
    );
    let html = reqwest::get(&url).await?.text().await?;
    let soup = Html::parse_document(&html);
    let code = soup
        .select(&Selector::parse("#program-source-text").expect("Failed to parse selector"))
        .next()
        .ok_or("Failed to find #program-source-text")?
        .text()
        .next()
        .ok_or("Failed to find text in #program-source-text")?;

    Ok(format!(
        "{} {} \\- {} \\- *{}*\n`{}`",
        contest_id,
        problem.index,
        markdown::escape(&problem.name),
        markdown::escape(verdict),
        markdown::escape_code(code)
    ))
}
